Launchpad: Add client-side handling for calypso_path task data

daledupreez opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/jetpack/pull/32177

## Proposed Changes

* This PR illustrates how we might take advantage of the proposed `calypso_path` task data returned from the server to simplify the client-side logic for tasks.
* The PR _does not_ remove any client-side handling for the time being, as we may have some tasks that aren't up to date, especially on Atomic sites, so it adds some logging via Tracks to identify tasks that don't have a path from the server and we either expect one to be present, or no client-side handler exists.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Ensure that your back end is running the changes from https://github.com/Automattic/jetpack/pull/32177
* Ensure you can inspect Tracks events
* Navigate to one or more sites that are currently showing a launchpad in Customer Home
* Verify that clicking on tasks that take you to other pages continue to work as expected
* Verify that you see one or more `calypso_launchpad_task_missing_calypso_path` Tracks events, and identify which tasks those are for -- in my testing that includes the newer `write_3_posts` and `enable_subscribers_modal ` tasks.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [N/A] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [N/A] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [N/A] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

daledupreez committed:
Launchpad: Add client-side handling for calypso_path task data

valterlorran commented on packages/launchpad/src/types.ts:
```diff
@@ -58,4 +59,5 @@ export interface LaunchpadTaskActionsProps {
 	extraActions?: {
 		setShareSiteModalIsOpen?: ( isOpen: boolean ) => void;
 	};
+	uiContext: 'calypso';
```
I think this should be optional. It's breaking some tests.
```suggestion
	uiContext?: 'calypso';
```
<img width="1003" alt="Screen Shot 2023-08-01 at 09 43 44" src="https://github.com/Automattic/wp-calypso/assets/1234758/b0368d77-1078-4334-95d9-6a003f7cb9df">

daledupreez replied:
I addressed this in [b283655](https://github.com/Automattic/wp-calypso/pull/80090/commits/b283655032172224d62cf5780bfdbbe849155901).

daledupreez committed:
Remove case for manage_subscribers

daledupreez committed:
Make uiContext optional

valterlorran reviewed (approved):
LGTM!

daledupreez committed:
Merge origin/trunk into add/calypso-path-data-handling-for-launchpad

daledupreez committed:
Reinstate custom handling and add logging for that code path

daledupreez commented:
Note that I've changed the implementation so we have the existing logic as a fallback, and added some logging so we know when tasks aren't supplying a `calypso_path` value. It's possible there is a better way to handle the logging, so please share if you have an alternative idea.

valterlorran reviewed (approved):
Still LGTM! Good idea to have a fallback!
